clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1647,19 +1645,57 @@ void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV) {
   auto RetVar = getOnly(returnVars);
   RetVar->mergeDeclaration(FromRetVar);
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->mergeDeclaration(FromVar);
+  handle_params(From, this,
+                [](ConstraintVariable *F, ConstraintVariable *T)
+                { T->mergeDeclaration(F); });
+}
+
+void FunctionVariableConstraint::handle_params
+(FVConstraint *From, FVConstraint *To,
+ std::function<void(ConstraintVariable*,ConstraintVariable*)> f) {
+  bool fromEmpty = From->numParams() == 0;
+  bool toEmpty = To->numParams() == 0;
+  bool paramsEq = From->numParams() == To->numParams();
+  assert(fromEmpty || toEmpty || paramsEq);
+  if (paramsEq) {
+    for (unsigned i = 0; i < From->numParams(); i++) {
+      CVarSet &FromP = From->getParamVar(i);
+      CVarSet &P = To->getParamVar(i);
+      auto FromVar = getOnly(FromP);
+      auto Var = getOnly(P);
+      f(FromVar, Var);
+    }
+  } else { // Dealing with an untyped prototype
+    FVConstraint *Empty = fromEmpty ? From : To;
+    FVConstraint *Typed = fromEmpty ? To : From;
+    for (auto deferred : Empty->getDeferredParams()) {
+      assert(Typed->numParams() == deferred.PS.size());
+      for(unsigned i = 0; i < deferred.PS.size(); i++) {
+        CVarSet ParamDC = Typed->getParamVar(i);
+        CVarSet ArgDC = deferred.PS[i];
+        ProgramInfo &Info = deferred.I;
```

Comment:
Per my comment above: You should just use the current `ProgramInfo` here. It's the same as the "deferred" one anyway, since these things are modified by reference.